core/src/common/empty_block/production.rs

```diff
@@ -0,0 +1,201 @@
+    parent_header: Block::Header,
+    provider_variant: ProviderVariant,
+    try_state: frame_try_runtime::TryStateSelect,
+) -> Result<((InherentData, Digest), Block::Header)>
+where
+    Block: BlockT<Hash = H256> + DeserializeOwned,
+    Block::Header: DeserializeOwned,
+    <Block::Hash as FromStr>::Err: Debug,
+    NumberFor<Block>: FromStr,
+    <NumberFor<Block> as FromStr>::Err: Debug,
+{
+    // We are saving state before we overwrite it while producing new block.
+    let backend = ext.as_backend();
+
+    log::info!(
+        "Producing new empty block at height {:?}",
+        *parent_header.number() + One::one()
+    );
+
+    let (next_block, new_block_building_info) = produce_next_block::<Block, HostFns>(
+        ext,
+        executor,
+        parent_header.clone(),
+        provider_variant,
+        previous_block_building_info,
+    )
+    .await?;
+
+    log::info!(
+        "Produced a new block ({})",
+        array_bytes::bytes2hex("0x", next_block.header().hash())
+    );
+
+    // And now we restore previous state.
+    ext.backend = backend;
+
+    pre_apply_inherents::<Block>(ext);
+    let state_root_check = true;
+    let signature_check = true;
+    let payload = (
+        next_block.clone(),
+        state_root_check,
+        signature_check,
+        try_state,
+    )
+        .encode();
+    call::<Block, _>(ext, executor, "TryRuntime_execute_block", &payload).await?;
+
+    log::info!("Executed the new block");
+
+    Ok((new_block_building_info, next_block.header().clone()))
+}
+
+/// Produces next block containing only inherents.
+pub async fn produce_next_block<Block: BlockT, HostFns: HostFunctions>(
+    externalities: &mut TestExternalities<HashingFor<Block>>,
+    executor: &WasmExecutor<HostFns>,
+    parent_header: Block::Header,
+    chain: ProviderVariant,
+    previous_block_building_info: Option<(InherentData, Digest)>,
+) -> Result<(Block, (InherentData, Digest))>
+where
+    Block: BlockT<Hash = H256> + DeserializeOwned,
+    Block::Header: DeserializeOwned,
+    <Block::Hash as FromStr>::Err: Debug,
+    NumberFor<Block>: FromStr,
+    <NumberFor<Block> as FromStr>::Err: Debug,
+{
+    let (inherent_data_provider, pre_digest) =
+        <ProviderVariant as InherentProvider<Block>>::get_inherent_providers_and_pre_digest(
+            &chain,
+            previous_block_building_info,
+            parent_header.clone(),
+            externalities,
+        )?;
+
+    pre_apply_inherents::<Block>(externalities);
+    let inherent_data = inherent_data_provider
+        .create_inherent_data()
+        .await
+        .map_err(|s| sc_cli::Error::Input(s.to_string()))?;
+    let digest = Digest { logs: pre_digest };
+
+    let header = Block::Header::new(
+        *parent_header.number() + One::one(),
+        Default::default(),
+        Default::default(),
+        parent_header.hash(),
+        digest.clone(),
+    );
+
+    call::<Block, _>(
```

Comment:
Maybe worth to check the version of the `Core` Runtime Interface to prevent weird future errors when we change something about it.

Reply:
I'm having trouble understanding how to decode the runtime api versions from this, creating a new issue to tackle it in.